### DISCUSSION TITLE
`save_hyperparameters()` is very slow when there are many hyperparameters, any speed up?

### DISCUSSION THREAD

**Discussion by VictorJouault** [Lightning-AI/pytorch-lightning#11281] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi,
I've currently refactored a part of code to use Pytorch Lightning instead of a regular pytorch script. I find a 35% speed up on a toy dataset, where the model had 20.9 K parameters.

However, when trying a bigger version of the model (9.1 M params), the initialization of the model takes a lot more time (even before lightning prints the number of params). I've managed to nail it down to the `self.save_hyperparameters()` function in init, since the initialization is instantaneous without calling `self.save_hyperparameters()`, but takes more than a full minute when calling it.

Removing `self.save_hyperparameters()` causes an issue in my code since I am calling  `model.load_from_checkpoint()` afterward. Would you have any thoughts on how I can speed up the code? The "regular" pytorch model manages to initialize, train and save the model faster than the Lightning one due to this slow down ...

Thanks!

**Comment by awaelchli** [comment]
What objects are you passing in? `self.save_hyperparameters()` is inspecting the signature, would be good to know which parameter is causing this.

**Comment by VictorJouault** [comment]
Thanks for answering! I'm simply calling `self.save_hyperparameters()` as part of the `__init__` function (code below). `model` is the main network model, which is a feedforward nn implemented with nn.module as base class. It contains an embedder and one input layer, 2 hidden layers and an output layer (in this case, the sizes of the layers are [6,000, 1008, 1008, 168]).

```
class EstimatorLightningModule(pl.LightningModule):
    def __init__(
        self,
        model: EstimatorNetwork,
        loss_scaling: Callable = None,
        lr: float = 1e-5,
        weight_decay: float = 1e-10,
    ) -> None:
        super().__init__()
        self.save_hyperparameters()
        self.model = model
        self.loss_scaling = loss_scaling
        self.lr = lr
        self.weight_decay = weight_decay
```

**Reply by awaelchli** [reply]
You have to set `self.save_hyperparameters(ignore="model")`, since you are passing in a full model as a "hyperparameter", you don't want to save that!

**Comment by VictorJouault** [comment]
Awesome thanks I’ll try this out! Just to make sure, this will still allow me to checkout the best model afterward? I’m unsure what you meant by the “signature” of the model. Thanks!

**Reply by awaelchli** [reply]
Yes, the docs explain it here: https://pytorch-lightning.readthedocs.io/en/latest/common/hyperparameters.html#lightningmodule-hyperparameters

Shows how to load the checkpoint as well. 


>  I’m unsure what you meant by the “signature” of the model.

This is the signature: 
``` 
def __init__(
        self,
        model: EstimatorNetwork,
        loss_scaling: Callable = None,
        lr: float = 1e-5,
        weight_decay: float = 1e-10,
    ) -> None:

```

**Comment by awaelchli** [comment, marked as answer]
You have to set` self.save_hyperparameters(ignore="model")`, since you are passing in a full model as a "hyperparameter", you don't want to save that!
